### PR TITLE
[charts/redis-ha] Removing PSP & Adding Kyverno adjustments to securityContexts. 

### LINKED ISSUE #204: [chart/redis-ha][REQUEST] Migrate from Pod Security Policy to SecurityContext
**Is your feature request related to a problem? Please describe.**

[Pod Security Policies](https://kubernetes.io/docs/concepts/security/pod-security-policy/) are deprecated and will be removed in Kubernetes v1.25.

Migration to another alternative is necessary.
To date there are some alternative solutions:
- [Pod Security Admission](https://kubernetes.io/docs/concepts/security/pod-security-admission/)
- [OPA Gatekeeper](https://github.com/open-policy-agent/gatekeeper)
- [Kyverno](https://github.com/kyverno/kyverno)
 
**Describe the solution you'd like**

A solution that is ecumenical and covers all the aforementioned alternatives is the use of [Security Context](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/) directly in the manifest of the chart.

This change must involve both ha-server and ha-proxy.

**Describe alternatives you've considered**

A possible solution would be to implement all the major alternatives within the chart but it is certainly a more onerous job.

**Additional context**

Related to #29

### PR DESCRIPTION
**Pull request by DandyDeveloper**
#### What this PR does / why we need it:
Adds the necessary adjustments to deprecate PSP and instead use appropriate securityContext options for containers/initContainers and tests. 

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #204

#### Special notes for your reviewer:
@joebowbeer @pierluigilenoci Please triple check and confirm you're happy with these changes. These are inline with the policy defined here: https://github.com/kyverno/policies//pod-security

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

File: charts/redis-ha/templates/redis-ha-statefulset.yaml [hunk 1]
```diff
@@ -263,6 +256,13 @@ spec:
         - redis-server
         args:
         - /data/conf/redis.conf
+        securityContext:
+          allowPrivilegeEscalation: false
+          capabilities:
+            drop:
+              - ALL
```
Comment (joebowbeer) [inline charts/redis-ha/templates/redis-ha-statefulset.yaml:263]: No indent sequences?

```suggestion
            - ALL
```

File: charts/redis-ha/values.yaml [hunk 2]
```diff
@@ -169,6 +161,16 @@ haproxy:
...
 
+  containerSecurityContext:
+    runAsNonRoot: true
+    allowPrivilegeEscalation: false
+    runAsUser: 1000
+    seccompProfile:
+      type: RuntimeDefault
+    capabilities:
+      drop:
+        - ALL
```
Comment (joebowbeer) [inline charts/redis-ha/values.yaml:172]:
```suggestion
      - ALL
```

No indent sequences?

File: charts/redis-ha/values.yaml [hunk 3]
```diff
@@ -388,6 +382,16 @@ securityContext:
...
 
+containerSecurityContext:
+  runAsNonRoot: true
+  allowPrivilegeEscalation: false
+  runAsUser: 1000
+  seccompProfile:
+    type: RuntimeDefault
+  capabilities:
+    drop:
+      - ALL
```
Comment (joebowbeer) [inline charts/redis-ha/values.yaml:393]:
```suggestion
    - ALL
```

**Review by joebowbeer — APPROVED** [review]
LGTM

**Review by joebowbeer — APPROVED** [review]
LGTM2

File: charts/redis-ha/templates/redis-ha-statefulset.yaml [hunk 4]
```diff
@@ -123,20 +122,15 @@ spec:
         {{- end }}
         command:
 {{ toYaml .Values.sysctlImage.command | indent 10 }}
-        securityContext:
-          runAsNonRoot: false
-          privileged: true
-          runAsUser: 0
```
Comment (34fathombelow) [inline charts/redis-ha/templates/redis-ha-statefulset.yaml:129]: Do we not need elevated privileges for initContainers with ```sysctlImage.enabled```?   I'm not very familiar with using this option but I can see certain task needing this.

File: charts/redis-ha/templates/redis-ha-statefulset.yaml [hunk 5]
```diff
@@ -123,20 +122,15 @@ spec:
...
+        securityContext: {{ toYaml .Values.containerSecurityContext | nindent 10 }}        
       {{- end }}
 {{- if and .Values.hostPath.path .Values.hostPath.chown }}
       - name: hostpath-chown
         image: {{ .Values.image.repository }}:{{ .Values.image.tag }}
-        securityContext:
-          runAsNonRoot: false
-          runAsUser: 0
```
Comment (34fathombelow) [inline charts/redis-ha/templates/redis-ha-statefulset.yaml:136]: Same as above.

File: charts/redis-ha/templates/redis-ha-statefulset.yaml [hunk 6]
```diff
@@ -238,9 +233,7 @@ spec:
            && if test -s /data/dump.rdb; \
               then cp -v /data/dump.rdb /data/dump.rdb_orig; fi \
            && mv -v /data/dump.rdb_ /data/dump.rdb"
-        securityContext:
-          runAsNonRoot: false
-          runAsUser: 0
```
Comment (34fathombelow) [inline charts/redis-ha/templates/redis-ha-statefulset.yaml:243]: Do we need elevated privileges here?

**Review by 34fathombelow — CHANGES_REQUESTED** [review]
Just a few things that caught my eye.  Thanks for all your hard work :)

File: charts/redis-ha/values.yaml [hunk 7]
```diff
@@ -63,6 +63,7 @@ serviceAccount:
 ## Recommend for externally exposed Redis clusters.
 ## ref: https://cbonte.github.io/haproxy-dconv/1.9/intro.html
 haproxy:
+  enabled: true
```
[the remark this answers is not in the document]
Comment (DandyDeveloper) [reply charts/redis-ha/values.yaml:66]: Oops, good catch thank you.

File: charts/redis-ha/values.yaml [hunk 8]
```diff
@@ -169,6 +161,16 @@ haproxy:
     fsGroup: 1000
     runAsNonRoot: true
 
+  containerSecurityContext:
+    runAsNonRoot: true
+    allowPrivilegeEscalation: false
+    runAsUser: 1000
```
[the remark this answers is not in the document]
Comment (DandyDeveloper) [reply charts/redis-ha/values.yaml:167]: Good catch.

File: charts/redis-ha/templates/redis-ha-statefulset.yaml (hunk 4 above)
Comment (DandyDeveloper) [reply charts/redis-ha/templates/redis-ha-statefulset.yaml:129]: I am going to remove this entirely. sysctl shouldn't really be used anymore this way.

File: charts/redis-ha/values.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (34fathombelow) [reply charts/redis-ha/values.yaml:66]: I also see you made this change in PR #233, not sure if it was from that.

[the remark this answers is not in the document]
Comment (DandyDeveloper) [reply charts/redis-ha/values.yaml:66]: It wasn't. I just have a habit of forgetting to disable this when I finish things.

**Comment by joebowbeer** [comment]
On second look, a couple of questions and a suggestion

Where is the Dockerfile for the default haproxy image?

The ArgoCD version runs as user 1000, but it runs the alpine-based image, which has a default nonroot user of 1000.

https://github.com/argoproj/argo-cd/blob/master/manifests/ha/base/redis-ha/chart/upstream.yaml#L1090

Suggestion: Remove the duplicated (pod) securityContext elements from the container securityContext values. That is, if the pod declares runsAsUser 1000 then remove the dup setting from the container.

File: charts/redis-ha/Chart.yaml [hunk 9]
```diff
@@ -1,12 +1,11 @@
 apiVersion: v2
 name: redis-ha
 home: http://redis.io/
-engine: gotpl
 keywords:
 - redis
 - keyvalue
 - database
-version: 4.19.1
+version: 4.21.0
```
Comment (pierluigilenoci) [inline charts/redis-ha/Chart.yaml:8]: Why does this version jump? What about 4.20?

**Review by pierluigilenoci — APPROVED** [review]
LGTM

File: charts/redis-ha/Chart.yaml (hunk 9 above)
Comment (DandyDeveloper) [reply charts/redis-ha/Chart.yaml:8]: I did 4.20 in a separate PR.

**Comment by DandyDeveloper** [comment]
@joebowbeer https://github.com/docker-library/haproxy/blob/master/Dockerfile.template#L13

UID 99 apparently. I'll apply and test with that adjustment. 

I'll remove the dupes on the containerSecurityContext in a sec.

**Comment by joebowbeer** [comment]
@DandyDeveloper wrote:
> UID 99 apparently. I'll remove the dupes on the containerSecurityContext in a sec.

If the different containers (redis vs haproxy) are running as different users then I suggest moving their user/group ids out of the PodSecurityContext and into their dedicated containerSecurityContext.

**Comment by DandyDeveloper** [comment]
@joebowbeer There's some need for duplicate fields between `securityContext` and `containerSecurityContext` because of the test pods using `containerSecurityContext`. 

Just FYI there are two `securityContext` fields in the values. One for HAProxy and another for Redis. 

As far as I can tell, this is "as good" as we'll get it without me spending unnecessary time on edge cases / conditionals in the test pods.

**Comment by joebowbeer** [comment]
> As far as I can tell, this is "as good" as we'll get it without me spending unnecessary time on edge cases / conditionals in the test pods.

np. I was thinking it would be better in some sense to move `{ runAsUser: 99, fsGroup: 99 }` into the container-specific context because this is the only part that actually changes per Dockerfile container, and it would avoid any confusion due to unintended inheritance. But really whatever works is great.

**Comment by DandyDeveloper** [comment]
@joebowbeer Yeah I 100% understand. Can you provide another review if you're happy to proceed?

**Comment by DandyDeveloper** [comment]
Uh... Woops.

**Review by joebowbeer — APPROVED** [review]
LGTM!
